### PR TITLE
[issue-553][issue-544]fix anyorder in schema with complex datatype

### PR DESCRIPTION
**Pull request by eason-lyx**
currently, we assume complex datatype should be at last.otherwise,exception occur.
for this pr:
1.fix any order in schema with complex datatype
2.modify some no correct parameter

### CODE REVIEW

**Comment by eason-lyx** [comment]
@ravikiran23 @nareshpr please help to review.

File: core/src/main/java/org/carbondata/core/carbon/metadata/schema/table/column/ColumnSchema.java [hunk 1]
```diff
@@ -105,6 +105,11 @@
   private boolean invisible = false;
 
   /**
+   * parent column schema
+   */
+  private ColumnSchema parentColumnSchema;
```
Comment (nareshpr) [inline core/src/main/java/org/carbondata/core/carbon/metadata/schema/table/column/ColumnSchema.java]: Why parentColumnSchema is required in childColumnSchema?
Thrift is not having parentColumnSchema place holder. Reloading table metadata to catalog may have some impact.

Better to handle columns reordering in CarbonSqlParser before calling Create after preparing table column list.

**Comment by QiangCai** [comment]
add links of issue
[issue-553](https://github.com/HuaweiBigData/carbondata/issues/553) Dataload need to sort order:dimension+complex+measure
[issue-544](https://github.com/HuaweiBigData/carbondata/issues/544) Data load failure when statement contain complex datatype and dictionary_include contain double type

**Comment by eason-lyx** [comment]
all comment handled.please review.
@nareshpr

**Comment by eason-lyx** [comment]
after discuss with @kumarvishal09 ,suggestion is when complex datatype is on at the end of dimension,then throw MalformedCarbonCommandException.

File: processing/src/main/java/org/carbondata/processing/graphgenerator/GraphGenerator.java [hunk 2]
```diff
@@ -67,7 +67,16 @@
 
 public class GraphGenerator {
 
+  /**
+   * global block info
+   */
   public static final HashMap<String, BlockDetails[]> blockInfo = new HashMap<>();
```
Comment (kumarvishal09) [inline processing/src/main/java/org/carbondata/processing/graphgenerator/GraphGenerator.java]: better use Map<String,BlockDetails[]> blockInfo = new HashMap<>()

File: integration/spark/src/main/scala/org/apache/spark/sql/CarbonSqlParser.scala [hunk 3]
```diff
@@ -473,6 +479,26 @@ class CarbonSqlParser()
   }
 
   /**
+   * check the dimensions whether complex type at last
+   * @param dims
+   * @return
+   */
+  protected def checkComplexDimPos(dims: Seq[Field]): Boolean = {
+    val complexDims: ArrayBuffer[(Field, Integer)] = new ArrayBuffer()
```
Comment (kumarvishal09) [inline integration/spark/src/main/scala/org/apache/spark/sql/CarbonSqlParser.scala]: only position you can store in the array buffer no need to store field

File: integration/spark/src/main/scala/org/apache/spark/sql/CarbonSqlParser.scala [hunk 4]
```diff
@@ -465,6 +465,12 @@ class CarbonSqlParser()
 
     val partitioner: Option[Partitioner] = getPartitionerObject(partitionCols, tableProperties)
 
+    if (!checkComplexDimPos(dims)) {
+      val errorMsg = "Complext data type column should be at the end of dimensions, " +
+        "please check the create statement."
+      throw new MalformedCarbonCommandException(errorMsg)
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/spark/sql/CarbonSqlParser.scala]: no need to define variable, please use directly

File: integration/spark/src/main/scala/org/apache/spark/sql/CarbonSqlParser.scala [hunk 5]
```diff
@@ -473,6 +479,26 @@ class CarbonSqlParser()
...
+  protected def checkComplexDimPos(dims: Seq[Field]): Boolean = {
+    val complexDims: ArrayBuffer[(Field, Integer)] = new ArrayBuffer()
+    val otherDims: ArrayBuffer[(Field, Integer)] = new ArrayBuffer()
+    dims.zipWithIndex.foreach { dim =>
+      dim._1.dataType.getOrElse("NIL") match {
+        case "array" =>
+          complexDims.+=((dim._1, dim._2))
+        case "struct" =>
+          complexDims.+=((dim._1, dim._2))
+        case _ => otherDims.+=((dim._1, dim._2))
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/spark/sql/CarbonSqlParser.scala]: suggest to remove point

**Comment by QiangCai** [comment]
better to find root cause and fix it.

**Comment by eason-lyx** [comment]
retest this please

**Comment by eason-lyx** [comment]
retest this please

**Pull request closed without merging**